File: server/CenterConnector/CenterConnector.cs
//////////////////////////////////////////////////////////////////////////
//
// 文件：server/CenterConnector/CenterConnector.cs
// 作者：Xoen Xie
// 时间：2023/06/15
// 描述：中心服连接器
// 说明：
//
//////////////////////////////////////////////////////////////////////////
#pragma warning disable CS8602, CS8600, CS8618

using XSF;
using Google.Protobuf.Collections;

namespace CC
{
    internal class CenterConnector : ICenterConnector
    {
        private Dictionary<uint, ServerInfo> m_ServerInfos;

        internal IServerInfoHandler? m_Handler;

        public CenterConnector()
        {
            m_ServerInfos = new Dictionary<uint, ServerInfo>();
        }

        public override bool Start()
        {
            var config = XSFCore.Config;

            return Connect(config.MainCenterIP, (int)config.CenterPort);
        }

        public override void DoRegist()
        {
            XSFCore.SetMessageExecutor((ushort)XsfPbid.SMSGID.CCcHandshake, new Executor_C_Cc_Handshake());
            XSFCore.SetMessageExecutor((ushort)XsfPbid.SMSGID.CCcServerInfo, new Executor_C_Cc_ServerInfo());
            XSFCore.SetMessageExecutor((ushort)XsfPbid.SMSGID.CCcServerLost, new Executor_C_Cc_ServerLost());
            XSFCore.SetMessageExecutor((ushort)XsfPbid.SMSGID.CCcServerOk, new Executor_C_Cc_ServerOk());
            XSFCore.SetMessageExecutor((ushort)XsfPbid.SMSGID.CCcStop, new Executor_C_Cc_Stop());
        }

        public override ModuleRunCode OnStartCheck()
        {
            if(IsHandshake)
            {
                return ModuleRunCode.OK;
            }

            return ModuleRunCode.Wait;
        }

        public override void OnOK()
        {
            var message = XSFCore.GetMessage((ushort)XsfPbid.SMSGID.CcCServerOk) as XsfMsg.MSG_Cc_C_ServerOk;
            message.mPB.ServerId = XSFCore.Server.ID;
            SendMessage(message);
        }

        public override void SendHandshake()
        {
            Serilog.Log.Information("CenterConnector SendHandshake");
            var message = XSFCore.GetMessage((ushort)XsfPbid.SMSGID.CcCHandshake) as XsfMsg.MSG_Cc_C_Handshake;
            message.mPB.ServerId = XSFCore.Server.ID;

            message.mPB.Ports.Clear();
            for(int i = 0; i < XSFCore.Server.Ports.Length; i ++)
            {
                message.mPB.Ports.Add(XSFCore.Server.Ports[i]);
            }

            SendMessage(message);
        }

        public override void SendHeartbeat()
        {
            //Serilog.Log.Information("CenterConnector SendHeartbeat");
            var message = XSFCore.GetMessage((ushort)XsfPbid.SMSGID.CcCHeartbeat);
            SendMessage(message);
        }

        public void AddInfo(XsfMsg.MSG_C_Cc_ServerInfo message)
        {
            for(int i = 0; i < message.mPB.Infos.Count; i ++)
            {
                bool IsNewAdd = false;
                var pbInfo = message.mPB.Infos[i];
                ServerInfo info = null;
                if(m_ServerInfos.TryGetValue(pbInfo.ServerId, out info))
                {
                    info.IP = pbInfo.Ip;
                    for(int J = 0; J < pbInfo.Ports.Count; J ++)
                    {
                        info.Ports[J] = pbInfo.Ports[J];
                    }
                    info.Status = (NodeStatus)pbInfo.Status;
                } 
                else
                {
                    IsNewAdd = true;
                    info = new ServerInfo();
                    info.ID = pbInfo.ServerId;
                    info.IP = pbInfo.Ip;
                    for(int J = 0; J < pbInfo.Ports.Count; J ++)
                    {
                        info.Ports[J] = pbInfo.Ports[J];
                    }
                    info.Status = (NodeStatus)pbInfo.Status;

                    m_ServerInfos.Add(info.ID, info);
                }

                var sid = ServerID.GetSID(info.ID);
                Serilog.Log.Information("【中心服连接器】收到服务器信息, id={0} [{1}-{2}-{3}] status={4}", info.ID, sid.ID, sid.Index, XSFCore.EP2CNName(sid.Type), info.Status);

                if(IsNewAdd)
                {
                    Serilog.Log.Information("【中心服连接器】新增服务器节点, id={0} [{1}-{2}-{3}] status={4}", info.ID, sid.ID, sid.Index, XSFCore.EP2CNName(sid.Type), info.Status);
                    if(m_Handler != null)
                        m_Handler.OnServerNew(info);

                    if(info.Status == NodeStatus.Ok)
                    {
                        Serilog.Log.Information("【中心服连接器】收到服务器节点已准备好, id={0} [{1}-{2}-{3}]", info.ID, sid.ID, sid.Index, XSFCore.EP2CNName(sid.Type) );
                        if(m_Handler != null)
                            m_Handler.OnServerOk(info);
                    }
                }
            }
        }

        public void OnNodeLost(uint nID)
        {
            m_ServerInfos.Remove(nID);
            var sid = ServerID.GetSID(nID);

            Serilog.Log.Information("【中心服连接器】有服务器节点离线, id={0} [{1}-{2}-{3}]", nID, sid.ID, sid.Index, XSFCore.EP2CNName(sid.Type));
            if(m_Handler != null)
                m_Handler.OnServerLost(nID);
        }

        public void OnNodeOk(uint nID)
        {
            ServerInfo info = null;
            if(m_ServerInfos.TryGetValue(nID, out info))
            {
                info.Status = NodeStatus.Ok;
                var sid = ServerID.GetSID(nID);

                Serilog.Log.Information("【中心服连接器】收到服务器节点已准备好, id={0} [{1}-{2}-{3}]", nID, sid.ID, sid.Index, XSFCore.EP2CNName(sid.Type) );

                if(m_Handler != null)
                    m_Handler.OnServerOk(info);
            }
            else
            {
                Serilog.Log.Error("服务器已准备好，但是本地未找到服务器信息, id={0}", nID);
            }
        }
    }
}

File: server-cpp/source/center/connector/src/CenterConnector.cpp
//////////////////////////////////////////////////////////////////////////
//
// 文件：center\connector\src\CenterConnector.cpp
// 作者：Xoen Xie
// 时间：2020/06/11
// 描述：中心服连接器
// 说明：
//
//////////////////////////////////////////////////////////////////////////

#include "CenterConnector.h"
#include "DMessageInc.h"
#include "CCExecutors.h"
#include "ICenterConnector.h"
 
namespace CC
{
    void CenterConnector::DoRegist()
    {
        XSFCore::SetMessageExecutor(xsf_pbid::C_Cc_Handshake, new Executor_C_Cc_Handshake());
        XSFCore::SetMessageExecutor(xsf_pbid::C_Cc_ServerInfo, new Executor_C_Cc_ServerInfo());
        XSFCore::SetMessageExecutor(xsf_pbid::C_Cc_ServerLost, new Executor_C_Cc_ServerLost());
        XSFCore::SetMessageExecutor(xsf_pbid::C_Cc_ServerOk, new Executor_C_Cc_ServerOk());
        XSFCore::SetMessageExecutor(xsf_pbid::C_Cc_Stop, new Executor_C_Cc_Stop());
    }

    bool CenterConnector::Start()
    {
        auto pConfig = XSFCore::GetServer()->GetConfig();

        return Connect(pConfig->MainCenterIP, pConfig->CenterPort);
    }

    void CenterConnector::Release(void)
    {
        XSF_DELETE(m_pHandler);
        
        NetConnector::Release();
    }

    uint8 CenterConnector::OnStartCheck() 
    {
        if(m_bHandshake)
        {
            return ModuleRunCode_OK;
        }

        return ModuleRunCode_Wait;
    }

    void CenterConnector::OnOK(void)
    {
        XSF_CAST(pMessage, XSFCore::GetMessage(xsf_pbid::Cc_C_ServerOk), xsf_msg::MSG_Cc_C_ServerOk);
        pMessage->mPB.set_server_id(XSFCore::GetServer()->GetSID()->ID);
        SendMessage(pMessage);
    }

    void CenterConnector::SendHandshake(void)
    {
        XSF_CAST(pMessage, XSFCore::GetMessage(xsf_pbid::Cc_C_Handshake), xsf_msg::MSG_Cc_C_Handshake);
        pMessage->mPB.set_server_id(XSFCore::GetServer()->GetSID()->ID);

        pMessage->mPB.clear_ports();

        auto pPorts = XSFCore::GetServer()->GetPorts();
        for(uint8 i = 0; i < EP_Max; i ++)
        {
            pMessage->mPB.add_ports(pPorts[i]);
        }

        SendMessage(pMessage);
    }

    void CenterConnector::SendHeartbeat(void)
    {
        SendMessage(XSFCore::GetMessage(xsf_pbid::Cc_C_Heartbeat));
    }

    void CenterConnector::AddInfo(xsf_msg::MSG_C_Cc_ServerInfo * pMessage)
    {
        for(int32 i = 0; i < pMessage->mPB.infos_size(); i ++)
        {
            auto & pbInfo = pMessage->mPB.infos(i);
            bool IsNewAdd = false;

            ServerInfo *pInfoFind = nullptr;
            auto it = m_Nodes.find(pbInfo.server_id());
            if(it != m_Nodes.end())
            {
                pInfoFind = it->second;
            }
            else
            {
                IsNewAdd = true;
                pInfoFind = new ServerInfo();
                pInfoFind->ID = pbInfo.server_id();
                m_Nodes.insert(InfoMap::value_type(pInfoFind->ID, pInfoFind));
            }

            strcpy(pInfoFind->sIP, pbInfo.ip().c_str());
            for(uint8 J = 0; J < EP_Max; J ++)
            {
                pInfoFind->Ports[J] = pbInfo.ports(J);
            }

            pInfoFind->Status = pbInfo.status();
            SID sid;
            sid.ID = pInfoFind->ID;
            XSF_INFO("【中心服连接器】收到服务器信息, id=%u [%u-%u-%s] status=%u", pInfoFind->ID, sid.S.server, sid.S.index, xsf::EP2CNName(sid.S.type), pInfoFind->Status);

            if(IsNewAdd)
            {
                XSF_INFO("【中心服连接器】新增服务器节点, id=%u [%u-%u-%s] status=%u", pInfoFind->ID, sid.S.server, sid.S.index, xsf::EP2CNName(sid.S.type), pInfoFind->Status);
                if(m_pHandler != nullptr)
                    m_pHandler->OnServerNew(pInfoFind);

                if(pInfoFind->Status == NodeStatus_Ok)
                {
                    XSF_INFO("【中心服连接器】收到服务器节点已准备好, id=%u [%u-%u-%s]", pInfoFind->ID, sid.S.server, sid.S.index, xsf::EP2CNName(sid.S.type));
                    if(m_pHandler != nullptr)
                        m_pHandler->OnServerOk(pInfoFind);
                }
            }
        }
    }

    void CenterConnector::OnNodeLost(uint32 nID)
    {
        auto it = m_Nodes.find(nID);
        if(it != m_Nodes.end())
        {
            delete it->second;
            m_Nodes.erase(it);
        }
        
        SID sid;
        sid.ID = nID;

        XSF_INFO("【中心服连接器】有服务器节点离线, id=%u [%u-%u-%s]", nID, sid.S.server, sid.S.index, xsf::EP2CNName(sid.S.type));
        if(m_pHandler != nullptr)
            m_pHandler->OnServerLost(nID);
    }

    void CenterConnector::OnNodeOk(uint32 nID)
    {
        auto it = m_Nodes.find(nID);
        if(it != m_Nodes.end())
        {
            it->second->Status = NodeStatus_Ok;

            SID sid;
            sid.ID = nID;

            XSF_INFO("【中心服连接器】收到服务器节点已准备好, id=%u [%u-%u-%s]", nID, sid.S.server, sid.S.index, xsf::EP2CNName(sid.S.type));
            if(m_pHandler != nullptr)
                m_pHandler->OnServerOk(it->second);
        }
        else
        {
            XSF_ERROR("服务器已准备好，但是本地未找到服务器信息, id=%u", nID);
        }
    }

    void SetCenterConnector(uint32 nModuleID, IServerInfoHandler *pInfoHandler)
    {
        NetConnectorInit * pInit = new NetConnectorInit();
        pInit->nModuleID = nModuleID;
        strcpy(pInit->sName, "CenterConnector");
        pInit->NeedReconnect = true;
        pInit->NoWaitStart = true;

        CenterConnector * pConnector = new CenterConnector();
        pConnector->SetHandler(pInfoHandler);
        XSFCore::GetServer()->AddModule(pConnector, pInit);
    }
}
